File: converters/TET-3.0-Linux/bind/perl/dumper.pl
# TET sample application for dumping PDF information with pCOS
#
# $Id: dumper.pl,v 1.3 2008/12/15 20:26:53 rjs Exp $
#/

use tetlib_pl 3.0;
use strict;

eval  {
    my $tet;

    if ($#ARGV != 0) {
	die("usage: dumper <filename>\n");
    }

    $tet = TET_new();

    my ($count, $pcosmode, $plainmetadata);
    my ($objtype, $i, $doc);
    my $docoptlist = "requiredmode=minimum";
    my $globaloptlist = "";
    # This is where input files live. Adjust as necessary.
    my $searchpath = "../data";
    my $optlist;

    TET_set_option($tet, $globaloptlist);

    $optlist = sprintf("searchpath={%s}", $searchpath);
    TET_set_option($tet, $optlist);


    if (($doc = TET_open_document($tet, $ARGV[0], $docoptlist)) == -1) {
	die(sprintf("ERROR: %s\n", TET_get_errmsg($tet)));
    }

    # --------- general information (always available) */

    $pcosmode = TET_pcos_get_number($tet, $doc, "pcosmode");

    printf("   File name: %s\n", TET_pcos_get_string($tet, $doc, "filename"));

    printf(" PDF version: %s\n",
	TET_pcos_get_string($tet, $doc, "pdfversionstring"));

    printf("  Encryption: %s\n",
	TET_pcos_get_string($tet, $doc, "encrypt/description"));

    printf("   Master pw: %s\n",
	TET_pcos_get_number($tet, $doc, "encrypt/master") ? "yes" : "no");

    printf("     User pw: %s\n",
	TET_pcos_get_number($tet, $doc, "encrypt/user") ? "yes" : "no");

    printf("Text copying: %s\n",
	TET_pcos_get_number($tet, $doc, "encrypt/nocopy") ? "no" : "yes");

    printf("  Linearized: %s\n",
	TET_pcos_get_number($tet, $doc, "linearized") ? "yes" : "no");

    if ($pcosmode == 0) {
	printf("Minimum mode: no more information available\n\n");
	TET_delete($tet);
	return 0;
    }

    # --------- more details (requires at least user password) */

    printf("PDF/X status: %s\n", TET_pcos_get_string($tet, $doc, "pdfx"));

    printf("PDF/A status: %s\n", TET_pcos_get_string($tet, $doc, "pdfa"));

    printf("  Tagged PDF: %s\n\n",
	TET_pcos_get_number($tet, $doc, "tagged") ? "yes" : "no");

    printf("No. of pages: %d\n",
	TET_pcos_get_number($tet, $doc, "length:pages"));

    printf(" Page 1 size: width=%g, height=%g\n",
	TET_pcos_get_number($tet, $doc, "pages[0]/width"),
	TET_pcos_get_number($tet, $doc, "pages[0]/height"));

    $count = TET_pcos_get_number($tet, $doc, "length:fonts");
    printf("No. of fonts: %d\n", $count);

    for ($i=0; $i < $count; $i++) {
	if (TET_pcos_get_number($tet, $doc, "fonts[$i]/embedded")) {
	    printf("embedded ");
	} else {
	    printf("unembedded ");
	}

	printf("%s font ",
	    TET_pcos_get_string($tet, $doc, "fonts[$i]/type"));
	printf("%s\n",
	    TET_pcos_get_string($tet, $doc, "fonts[$i]/name"));
    }

    printf("\n");

    $plainmetadata =
	TET_pcos_get_number($tet, $doc, "encrypt/plainmetadata");

    if ($pcosmode == 1 && !$plainmetadata &&
	    (TET_pcos_get_number($tet, $doc, "encrypt/nocopy"))) {
	printf("Restricted mode: no more information available\n\n");
	TET_delete($tet);
	return 0;
    }

    # --------- document info keys and XMP metadata (requires master pw
    # or plaintext metadata)
    #/

    $count = TET_pcos_get_number($tet, $doc, "length:/Info");

    for ($i=0; $i < $count; $i++) {
	$objtype = TET_pcos_get_string($tet, $doc, "type:/Info[$i]");
	printf("%12s: ", TET_pcos_get_string($tet, $doc, "/Info[$i].key"));

	# Info entries can be stored as string or name objects */
	if ($objtype eq "string" || $objtype eq "name") {
	    printf("'%10s'\n",
		TET_pcos_get_string($tet, $doc, "/Info[$i]"));
	} else {
	    printf("(%s object)\n",
		TET_pcos_get_string($tet, $doc, "type:/Info[$i]"));
	}
    }

    printf("\nXMP meta data: ");

    $objtype = TET_pcos_get_string($tet, $doc, "type:/Root/Metadata");
    if ($objtype eq "stream") {
	my $contents;

	$contents =
	    TET_pcos_get_stream($tet, $doc, "", "/Root/Metadata");
	printf("%d bytes ", length($contents));

	# This demonstrates Unicode conversion */
	my $dummy = TET_utf8_to_utf16($tet, $contents, "utf16");
	printf("(%d Unicode characters)\n\n", length($dummy)/2);
    } else {
	printf("not present\n\n");
    }

    TET_close_document($tet, $doc);
};

if ($@) {
    printf("TET exception occurred in dumper:\n");
    printf("$@\n");
    exit;
}

File: converters/TET-3.0-Linux/bind/python/dumper.py
# TET sample application for dumping PDF information with pCOS
#
# $Id: dumper.py,v 1.6 2008/12/15 20:26:54 rjs Exp $
#

import sys
from tetlib_py import *

def yesno(arg):
    if (arg != 0):
	return "yes"
    return "no"

if len(sys.argv) != 2:
    raise Exception("usage: dumper <filename>")

try:
    try:
	tet = TET_new()

	searchpath = "searchpath={../data}"
	docoptlist = "requiredmode=minimum"
	globaloptlist = ""

	TET_set_option(tet, searchpath)
	TET_set_option(tet, globaloptlist)

	doc = TET_open_document(tet, sys.argv[1], docoptlist)
	if doc  == -1:
	    raise Exception("ERROR: %s\n" % TET_get_errmsg(tet))

	# --------- general information (always available) */
	pcosmode = int(TET_pcos_get_number(tet, doc, "pcosmode"))

	print("   File name: %s" % TET_pcos_get_string(tet, doc, "filename"))

	print(" PDF version: %s" % \
	    TET_pcos_get_string(tet, doc, "pdfversionstring"))

	print("  Encryption: %s" % \
	    TET_pcos_get_string(tet, doc, "encrypt/description"))

	print("   Master pw: %s" % \
	    yesno(TET_pcos_get_number(tet, doc, "encrypt/master")))

	print("     User pw: %s" % \
	    yesno(TET_pcos_get_number(tet, doc, "encrypt/user")))

	print("Text copying: %s" % \
	    yesno(not TET_pcos_get_number(tet, doc, "encrypt/nocopy")))

	print("  Linearized: %s" % \
	    yesno(TET_pcos_get_number(tet, doc, "linearized")))

	if (pcosmode == 0):
	    print("Minimum mode: no more information available\n")
	    TET_close_document(tet, doc)
	    exit(0)

	# --------- more details (requires at least user password) */

	print("PDF/X status: %s" % TET_pcos_get_string(tet, doc, "pdfx"))

	print("PDF/A status: %s" % TET_pcos_get_string(tet, doc, "pdfa"))

	print("  Tagged PDF: %s\n" % \
	    yesno(TET_pcos_get_number(tet, doc, "tagged")))

	print("No. of pages: %d" % \
	    int(TET_pcos_get_number(tet, doc, "length:pages")))

	print(" Page 1 size: width=%g, height=%g" % \
	    (TET_pcos_get_number(tet, doc, "pages[%d]/width" % 0),
	     TET_pcos_get_number(tet, doc, "pages[%d]/height" % 0)))

	count = int(TET_pcos_get_number(tet, doc, "length:fonts"))
	print("No. of fonts: %d" % count)

	for i in range(count):
	    if (TET_pcos_get_number(tet, doc, "fonts[%d]/embedded" % i)):
		print "embedded",
	    else:
		print "unembedded",

	    print("%s font" % \
		TET_pcos_get_string(tet, doc, "fonts[%d]/type" % i)),
	    print("%s" % \
		TET_pcos_get_string(tet, doc, "fonts[%d]/name" % i))

	print

	plainmetadata = \
		int(TET_pcos_get_number(tet, doc, "encrypt/plainmetadata"))

	if (pcosmode == 1 and not plainmetadata and \
		int(TET_pcos_get_number(tet, doc, "encrypt/nocopy"))):
	    print("Restricted mode: no more information available\n")
	    TET_close_document(tet, doc)
	    exit(0)

	# --------- document info keys and XMP metadata (requires master pw
	# or plaintext metadata)
	#

	count = int(TET_pcos_get_number(tet, doc, "length:/Info"))

	for i in range(0, count):
	    objtype = int(TET_pcos_get_number(tet, doc, "type:/Info[%d]" % i))
	    print("%12s:" % TET_pcos_get_string(tet,doc, "/Info[%d].key" % i)),

	    # Info entries can be stored as string or name objects */
	    # pcos_to_sting == 4; pcos_ot_name == 3
	    if (objtype == 4 or objtype == 3):
		print("'%10s'" % \
		    TET_pcos_get_string(tet, doc, "/Info[%d]" % i))
	    else:
		print("(%s object)" % \
		    TET_pcos_get_string(tet, doc, "type:/Info[%d]" % i))

	print("\nXMP meta data:"),

	objtype = int(TET_pcos_get_number(tet, doc, "type:/Root/Metadata"))
	# pcos_ot_stream == 7
	if (objtype == 7):
	    contents = TET_pcos_get_stream(tet, doc, "", "/Root/Metadata")
	    print("%d bytes" % len(contents)),

	    ustring = unicode(contents, "UTF-8")
	    print("(%d Unicode characters)\n" % len(ustring))
	else:
	    print("not present\n")

	TET_close_document(tet, doc)

    except TETlibException:
	print("\nError " + repr(TET_get_errnum(tet)) + "in " + \
		TET_get_apiname(tet) + ": " + TET_get_errmsg(tet) + "\n")

    except Exception, e:
	print "Exception occurred:"
	print e

finally:
    TET_delete(tet)
